hbase-client/src/main/java/org/apache/hadoop/hbase/client/BufferedMutatorThreadPoolExecutor.java

```diff
@@ -0,0 +1,197 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hbase.client;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.util.Threads;
+
+import java.lang.reflect.Field;
+import java.util.Map;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicLong;
+
+@SuppressWarnings("WeakerAccess")
+public class BufferedMutatorThreadPoolExecutor extends ThreadPoolExecutor {
```

Comment:
OK so the key here is this customized thread pool. Please add more comments to say what is going on here? And how do we make use of this pool? Haven't seen where we create this pool or call getPoolExecutor?

Reply:
It's available for the application. I leave the default pool intact ...

```
        BufferedMutatorThreadPoolExecutor pool = BufferedMutatorThreadPoolExecutor.getPoolExecutor(hadoopConf);

        BufferedMutatorParams mutatorParms = new BufferedMutatorParams(tableName)
            .listener(listener)
            .pool(pool);
```

Reply:
So without using this pool, can we still get the 10x throughput? Or it will still get a better performance but not 10x?

Reply:
Exactly. This class can be suppressed (it is used), it is just an illustration on how to get some statistics on flush().
The best way to achieve that, would be that the Future contains all needed information.

Reply:
What's next? Is there something blocking the review?